---
title: Gaussian Mixture Models in R
abstract: ' Gaussian mixture models (GMMs) are widely used for modelling stochastic
  problems. Indeed, a wide diversity of packages have been developed in R. However,
  no recent review describing the main features offered by these packages and comparing
  their performances has been performed. In this article, we first introduce GMMs
  and the EM algorithm used to retrieve the parameters of the model and analyse the
  main features implemented among seven of the most widely used R packages. We then
  empirically compare their statistical and computational performances in relation
  with the choice of the initialisation algorithm and the complexity of the mixture.
  We demonstrate that the best estimation with well-separated components or with a
  small number of components with distinguishable modes is obtained with REBMIX initialisation,
  implemented in the [rebmix](https://CRAN.R-project.org/package=rebmix) package,
  while the best estimation with highly overlapping components is obtained with *k*-means
  or random initialisation. Importantly, we show that implementation details in the
  EM algorithm yield differences in the parameters'' estimation. Especially, packages
  [mixtools](https://CRAN.R-project.org/package=mixtools) (Young et al. 2020) and
  [Rmixmod](https://CRAN.R-project.org/package=Rmixmod) (Langrognet et al. 2021) estimate
  the parameters of the mixture with smaller bias, while the RMSE and variability
  of the estimates is smaller with packages [bgmm](https://CRAN.R-project.org/package=bgmm)
  (Ewa Szczurek 2021) , [EMCluster](https://CRAN.R-project.org/package=EMCluster)
  (W.-C. Chen and Maitra 2022) , [GMKMcharlie](https://CRAN.R-project.org/package=GMKMcharlie)
  (Liu 2021), [flexmix](https://CRAN.R-project.org/package=flexmix) (Gruen and Leisch
  2022) and [mclust](https://CRAN.R-project.org/package=mclust) (Fraley, Raftery,
  and Scrucca 2022). The comparison of these packages provides R users with useful
  recommendations for improving the computational and statistical performance of their
  clustering and for identifying common deficiencies. Additionally, we propose several
  improvements in the development of a future, unified mixture model package. '
author:
- name: Bastien Chassagnol
  affiliation: |-
    Laboratoire de Probabilités, Statistiques et Modélisation (LPSM), UMR
    CNRS 8001
  address:
  - 4 Place Jussieu Sorbonne Université
  - 75005, Paris, France
  - '*ORCiD: [0000-0002-8955-2391](https://orcid.org/0000-0002-8955-2391)*'
  - |
    [`bastien_chassagnol@laposte.net`](mailto:bastien_chassagnol@laposte.net)
- name: Antoine Bichat
  affiliation: Les Laboratoires Servier
  address:
  - 50 Rue Carnot
  - 92150, Suresnes, France
  - <https://rdrr.io/github/abichat/abutils/>
  - '*ORCiD: [0000-0001-6599-7081](https://orcid.org/0000-0001-6599-7081)*'
  - |
    [`antoine.bichat@servier.com`](mailto:antoine.bichat@servier.com)
- name: Cheïma Boudjeniba
  affiliation: |-
    Systems Biology Group, Dept. of Computational Biology, Institut
    Pasteur
  address:
  - 25 Rue du Dr Roux
  - 75015 Paris
  - |
    [`cheima.boudjeniba@servier.com`](mailto:cheima.boudjeniba@servier.com)
- name: Pierre-Henri Wuillemin
  affiliation: Laboratoire d'Informatique de Paris 6 (LIP6), UMR 7606
  address:
  - 4 Place Jussieu Sorbonne Université
  - 75005, Paris, France
  - <http://www-desir.lip6.fr/~phw/>
  - '*ORCiD: [0000-0003-3691-4886](https://orcid.org/0000-0003-3691-4886)*'
  - |
    [`pierre-henri.wuillemin@lip6.fr`](mailto:pierre-henri.wuillemin@lip6.fr)
- name: Mickaël Guedj
  affiliation: Les Laboratoires Servier
  address:
  - 50 Rue Carnot
  - 92150, Suresnes, France
  - <https://michaelguedj.github.io/>
  - '*ORCiD: [0000-0001-6694-0554](https://orcid.org/0000-0001-6694-0554)*'
  - |
    [`mickael.guedj@gmail.com`](mailto:mickael.guedj@gmail.com)
- name: David Gohel
  affiliation: ArData
  address:
  - 59 rue Voltaire
  - 92800, PUTEAUX, France
  - <https://www.ardata.fr/expertise-r/>
  - '*ORCiD: [0000-0003-2837-8884](https://orcid.org/0000-0003-2837-8884)*'
  - |
    [`david.gohel@ardata.fr`](mailto:david.gohel@ardata.fr)
- name: Gregory Nuel
  affiliation: |-
    Laboratoire de Probabilités, Statistiques et Modélisation (LPSM), UMR
    CNRS 8001
  address:
  - 4 Place Jussieu Sorbonne Université
  - 75005, Paris, France
  - <http://nuel.perso.math.cnrs.fr/>
  - '*ORCiD: [0000-0001-9910-2354](https://orcid.org/0000-0001-9910-2354)*'
  - |
    [`Gregory.Nuel@math.cnrs.fr`](mailto:Gregory.Nuel@math.cnrs.fr)
- name: Etienne Becht
  affiliation: Les Laboratoires Servier
  address:
  - 50 Rue Carnot
  - 92150, Suresnes, France
  - '*ORCiD: [0000-0003-1859-9202](https://orcid.org/0000-0003-1859-9202)*'
  - |
    [`etienne.becht@servier.com`](mailto:etienne.becht@servier.com)
date: '2023-11-01'
date_received: '2021-12-06'
journal:
  firstpage: 56
  lastpage: 76
volume: 15
issue: 2
slug: RJ-2023-043
packages:
  cran: ~
  bioc: ~
draft: no
preview: preview.png
bibliography: chassagnol-becht-nuel-benchmark-of-Gaussian-mixtures.bib
CTV: ~
preamble: |
  \usepackage[most]{tcolorbox}
  \usepackage{booktabs}
  \usepackage{bm, amsmath}
  \usepackage{array}
  \usepackage{multirow}
  \usepackage{wrapfig}
  \usepackage{float}
  \usepackage{colortbl}
  \usepackage{xcolor}
  \usepackage{longtable}
  \usepackage{tabu}
  \usepackage{adjustbox}
  \usepackage{threeparttable}
  \usepackage{makecell}
  \def\greentick{\includegraphics[scale=0.05]{figures/green_tick.png}}
  \def\redcross{\includegraphics[scale=0.05]{figures/red_cross.png}}
  \newtcolorbox{blackbox}[1]{colback=white, colframe=black, coltext=black, boxsep=1.5pt, arc=4pt, before=\centering, title=#1}
  \newenvironment{cols}[1][]{}{} \newenvironment{col}[1]{\begin{minipage}{#1}\ignorespaces}{\end{minipage} \ifhmode\unskip\fi
  \aftergroup\useignorespacesandallpars} \def\useignorespacesandallpars#1\ignorespaces\fi{#1\fi\ignorespacesandallpars}
output:
  rjtools::rjournal_web_article:
    self_contained: no
    toc: no
    legacy_pdf: no

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  fig.align = "center",
  fig.fullwidth = TRUE,
  message = FALSE,
  warning = FALSE,
  out.width = "100%",
  cache = FALSE, lazy.cache = FALSE
)

library(ggplot2)
library(dplyr)
library(flextable)
library(kableExtra)
library(downlit)
library(RGMMBench)

options(dplyr.summarise.inform = FALSE) # remove the override .groups argument

# rjtools::initial_check_article(".", dic = "en_GB", ignore=c("gmms", "transcriptomic", "gaussian", "heteroscedascity", "equi", "clusterwise", "underbrace", "homoscedascity", "rebmix", "cran", "bioconductor", "mclust", "kd", "st", "th", "homoscedastic", "hyperparameters"))
```


```{css, echo=FALSE}
p.caption {
  font-size: 0.6em;
}

.blackbox {
  padding: 1em;
  background: #DBDBDB;
  color: black;
  border: 4px solid black;
  border-radius: 10px;
}
.center {
  text-align: center;
}

.cols {display: flex; }
```

# Introduction to Mixture modelling

Formally, let's consider a pair of random variables $(X,S)$ with $S \in \{1, \ldots, k\}$ a discrete variable and designing
the component identity of each observation. When observed, $S$ is
generally denoted as the labels of the individual observations. $k$ is
the number of mixture *components*. Then, the density distribution of
$X$ is given in Equation \@ref(eq:1):

```{=tex}
\begin{equation}
\begin{split}
f_\theta(X) &= \sum_S f_\theta (X, S) \\
&= \sum_{j=1}^k p_j f_{\zeta j}(X), \quad X\in\mathbb{R}
\end{split}
(\#eq:1)
\end{equation}
```
where $\theta = (p, \zeta) = (p_1, \ldots, p_k, \zeta_1, \ldots, \zeta_k)$
denotes the parameters of the model: $p_j$ is the proportion of
component $j$ and $\zeta_j$ represents the parameters of the density
distribution followed by component $j$. In addition, since $S$ is a categorical variable parametrized by $p$, the prior weights must enforce the unit simplex constraint (Equation \@ref(eq:2)):


```{=tex}
\begin{equation}
\begin{cases}
p_j \ge 0 \quad \forall j \in \{1, \ldots, k \}\\
\sum_{j=1}^k p_j =1
\end{cases}
(\#eq:2)
\end{equation}
```
In terms of applications, mixture models can be used to achieve the
following goals:

-   *Clustering*: hard clustering consists in determining a complete
    partition of the $n$ observations $x_{1:n}$ into $k$ disjoint
    non-empty subsets. In the context of *mixture model-based
    clustering*, this is done by assigning each observation $i$ to the
    cluster $\hat{s_i}=\arg \max_j \eta_{i} (j)$ that maximises the
    posterior distribution (MAP) (see Equation
    \@ref(eq:posteriori)):

```{=tex}
\begin{equation}
        \eta_{i} (j) := \mathbb{P}_{\theta} (S_i=j |X_i=x_i)
    (\#eq:posteriori)
\end{equation}
```
-   *Prediction*: the purpose is to predict a response variable $Y$ from
    an explanatory variable $X$. The dependent variable $Y$ can either
    be discrete, taking values in classes $\{1, \ldots, G\}$
    (*classification* task) or continuous (*regression* task). In this
    paper, we do not extensively discuss application of mixture models to regression purposes but refer the reader
    to @bouveyron_girard09 for mixture classification and
    @shimizu_kaneko20 for mixtures of regression models.

In section [Univariate and multivariate Gaussian distributions in the context of mixture models], we describe the most
commonly used family, the Gaussian Mixture Model (GMM). We then present
the MLE estimation of the parameters of a GMM, introducing the classic
EM algorithm in section [Parameter estimation in finite mixtures models]. Finally, we introduce bootstrap methods used to evaluate the
quality of the estimation and metrics used for the selection of the best
model in respectively appendices *Derivation of confidence intervals in GMMs* and *Model selection*.


## Univariate and multivariate Gaussian distributions in the context of mixture models

We focus our study on the finite Gaussian mixture models (GMM) in which we
suppose that each of the $k$ components follows a Gaussian
distribution.



We recall below the definition of the Gaussian
distribution in both univariate and multivariate context. In the finite univariate Gaussian mixture model, the distribution of
each component $f_{\zeta j}(X)$ is given by the following univariate
Gaussian p.d.f. (probability density function) (Equation
\@ref(eq:gaussian-dist)):

\begin{equation}
f_{\zeta j}(X=x)=\varphi_{\zeta_j}(x | \mu_j, \sigma_j):=\frac{1}{\sqrt{2\pi} \sigma_j} \exp^{- \frac{(x - \mu_j)^2}{2 \sigma_j^2}}
(\#eq:gaussian-dist)
\end{equation} which we note: $X \sim \mathcal{N}(\mu_j, \sigma_j)$.

In the univariate case, the parameters to be inferred from each
component, $\zeta_j$, are: $\mu_j$, the *location* parameter (equal to
the mean of the distribution) and $\sigma_j$, the *scale* parameter
(equal to the standard deviation of the distribution with a Gaussian distribution).

Following parsimonious parametrisations with respect to univariate GMMs
are often considered:

-   *homoscedascity*: variance is considered equal for all components,
    $\sigma_j = \sigma, \forall j \in \{1, \ldots, k \}$, as opposed to
    heteroscedascity where each sub-population has its unique
    variability.

-   *equi-proportion* among all mixtures:
    $p_j = \frac{1}{k} \quad j \in \{ 1, \ldots, k\}$ [^1]

[^1]: A rarer constraint considered implies to enforce a linear
    constraint over the clusters' means, of the following general form:
    $\sum_{j=1}^k a_j \mu_j=0$, with $\{a_1, \ldots, a_k\}$. For
    instance, the R package \pkg{epigenomix} considers a $k=3$ component
    mixture in the context of transcriptomic (differential analyses) and
    epigenetic (histone modification) to automatically identify
    undifferentiated, over and under-expressed genes between case and
    control samples. A common constraint then is to enforce the
    distribution of fold changes corresponding to the undifferentiated
    expressed genes to have a distribution centred on 0. Combining
    equality of means and equality of variances is irrelevant, as the
    model is then degenerate. Additionally, setting constraints on the
    means makes the estimation of the parameters challenging, as
    detailed in Appendix *Extensions of the EM algorithm to overcome its limitations*.


In the finite multivariate Gaussian mixture model, the distribution $f_{\zeta j}(\boldsymbol{X})$ of each component $j$, where
$\boldsymbol{X} \in \mathbb{R}^D =(X_1, \ldots, X_D)^\top$ is a multivariate random variable
of dimension $D$, is given by the
following multivariate Gaussian p.d.f.
(Equation \@ref(eq:multivariate-distribution)):

```{=tex}
\begin{equation}
    f_{\zeta j}(\boldsymbol{X}=\boldsymbol{x})=\operatorname{det}(2\pi\boldsymbol{\Sigma}_j)^{-\frac{1}{2}} \exp\left( -\frac{1}{2} (\boldsymbol{x} - \boldsymbol{\mu}_j) \boldsymbol{\Sigma}_j^{-1} (\boldsymbol{x} - \boldsymbol{\mu}_j)^\top\right)
(\#eq:multivariate-distribution)
\end{equation}
```
which we note
$\boldsymbol{X} \sim \mathcal{N}_D(\boldsymbol{\mu}_j, \boldsymbol{\Sigma}_j)$. The parameters to be estimated for each component can be decomposed into:

-   $\boldsymbol{\mu}_j=\begin{pmatrix} \mu_{1j} \\ \vdots \\ \mu_{Dj} \end{pmatrix} \in \mathbb{R}^D$, the $D$-dimensional mean vector.

-   $\boldsymbol{\Sigma}_j$, the $\mathcal{M}_D(\mathbb{R})$ positive-definite [^3]covariance matrix, whose diagonal terms are the individual variances of each feature and the off-diagonal terms are the pairwise covariance terms.

[^3]: The positive-definiteness constraint can be interpreted from a probabilistic point of view as a necessary condition such that the generalised integral of the multivariate distribution is defined and sum-to-one over $\mathbb{R}$ or from the statistical definition of the covariance. A symmetric real matrix $\boldsymbol{X}$ of rank $D$ is said to be *positive-definite* if for any non-zero vector
$\mathbf{v}, \in \mathbb{R}^D$, the following constraint
$\mathbf{v}^\top \boldsymbol{X} \mathbf{v} > 0$ is enforced.

Three families of multivariate GMMs are often considered:

-   the *spherical* family, $\boldsymbol{\Sigma}_j=\sigma_j^2 \boldsymbol{I}_D$, with $\sigma_j \in \mathbb{R}_{+}^*$, refers to GMMs whose covariance matrix is diagonal with an unique standard deviation term.  The corresponding volume representation is a $D-$*hypersphere* of radius $\sigma_j$.
-   the *diagonal* family, $\boldsymbol{\Sigma}_j=\operatorname{diag} \left(\sigma_{1j}^2, \ldots, \sigma_{1D}^2\right)$, with $\sigma_j \in \mathbb{R}_{+}^D$, refers to GMMs whose covariance matrix is diagonal. Its associated volume representation is an ellipsoid whose main axes are aligned with the $D$ canonical basis of $\mathbb{R}^D$. Of note, the null constraint imposed over the off-diagonal terms in the spherical and diagonal families imply that the multivariate distribution can be further decomposed and analysed as the product of univariate independent Gaussian realisations.
-   the *ellipsoidal* family, also named the *general* family, refer to GMMs whose covariance matrix, $\boldsymbol{\Sigma}_j$, can be any arbitrary positive-definite $D \times D$ matrix. Thus, the corresponding clusters for each component $J$ are ellipsoidal, centred at the mean vector $\boldsymbol{\mu}_j$, and volume and orientation respectively determined by the eigenvalues and the eigenvectors of the covariance matrix $\boldsymbol{\Sigma}_j$.


In the multivariate setting, the volume, shape, and orientation of the covariances can be constrained to be equal or variable across clusters, generating 14 possible parametrisations with different geometric characteristics [@banfield_raftery93; @celeux_govaert92]. We review them in Appendix *Parameters estimation in a high-dimensional context* and Table \@ref(tab:parameter-configuration-bivariate). Of note, the
correlation matrix can be easily derived from the covariance
matrix with the following normalisation:
$\operatorname{cor}(\boldsymbol{X})=\left(\frac{\operatorname{cov}(x_l, x_m)}{\sqrt{\operatorname{var}(x_l)} \times \sqrt{\operatorname{var}(x_m)}}\right)_{(l,m) \in D \times D}$. Correlation if strictly included between -1 and 1, the strength of the
correlation is given by its absolute value while the type of the
interaction is returned by its sign. A correlation of 1 or -1 between two features indicates a strictly linear relationship.


For the sake of simplicity and tractability, we will only consider the
fully unconstrained model in both the univariate (heteroscedastic and
unbalanced classes) and multivariate dimension (unbalanced and complete
covariance matrices for each cluster) in the remainder of our paper.


## Parameter estimation in finite mixtures models

A common way for estimating the parameters of a parametric distribution is
the *maximum likelihood estimation* (MLE) method. It consists in
estimating the parameters by maximising the likelihood, or
equivalently the log-likelihood of a sample. In what follows,
$\ell(\theta|x_{1:n})=\log (f(x_{1:n}|\theta))$ is the log-likelihood of
a *n*-sample. When all observations are independent, it simplifies to
$\ell(\theta|x_{1:n}) = \sum_{i=1}^n \log (f(x_i|\theta))$. The MLE
consists in finding the parameter estimate $\hat{\theta}$ which maximises the
log-likelihood $\hat{\theta} = \arg \max \ell (\theta | x_{1:n})$.

Recovering the maximum of a function is generally performed by
finding the values at which its derivative vanishes. The MLE in GMMs
has interesting properties, as opposed to the *moment estimation*
method: it is a consistent, asymptotically efficient and unbiased
estimator [@chen16; @mclachlan_peel00].

When $S$ is completely observed, for pairs of observations
$(x_{1:n}, s_{1:n})$, the log-likelihood of a finite mixture model is
simply given by Equation \@ref(eq:6):

```{=tex}
\begin{equation}
\ell(\theta|X_{1:n}=x_{1:n}, S_{1:n}=s_{1:n})=\sum_{i=1}^n \sum_{j=1}^k \left[\log\left(f_{\zeta_j} (x_i, s_i=j)\right) + \log(p_j) \right]_{\mathbf{1}_{s_i=j}}
(\#eq:6)
\end{equation}
```
where an analytical solution can be computed provided that a closed-form estimate exists to retrieve the parameters $\zeta_j$ for each components' parametric distribution. The MLE maximisation, in this context, involves the estimation of the parameters for each cluster, denoted as $\zeta_j$. The corresponding proportions, $p_j$, can be straightforwardly computed as the ratios of observations assigned to cluster $j$ relative to the total number of observations, $n$.

However, when $S$ is unobserved, the log-likelihood, qualified as
incomplete with respect to the previous case, is given by Equation
\@ref(eq:7):

```{=tex}
\begin{equation}
\ell (\theta \vert x_{1:n}) = \sum_{i=1}^n  \log \left( \underbrace{\sum_{j=1}^k  p_j f_{\zeta_j}(x_i)}_{\text{sum of of logs}} \right)
(\#eq:7)
\end{equation}
```
The sum of terms embed in the log function (see underbrace section in Equation \@ref(eq:7)) makes it intractable in practice to derive the null values of its corresponding derivative. Thus, no closed form of the MLE is available,
including for the basic univariate GMM model. This is why most
parameter estimation methods derive instead from the *EM algorithm*,
first described in @dempster_etal77. We describe its main theoretical properties, the reasons for its popularity, and its main limitations in the next section.

## The EM algorithm
In cases where both $S$ and the parameters associated to each cluster are unknown, there is no available closed-form solution that would jointly maximise the log-likelihood, as defined in Equation \@ref(eq:7), with respect to the set of parameters $({\theta}, S)$.
However, when either $S$ or $\theta$ are known, the estimation of the
other parameters is straightforward. Hence, the general principle of
EM-like algorithms is splitting this complex non-closed joint MLE
estimation of $(S, \theta)$ into the iterative estimation of $S_q$ from
$\hat{\theta}_{q-1}$ and $X$ (expectation phase, or *E-step* of the algorithm)
and the estimation of $\hat{\theta}_{q}$ from $(S_q$ and $X$ (maximisation phase, or *M-step*), with $\hat{\theta}_{q-1}$ being the estimated parameters at the
previous step $q-1$, until we reach the convergence.


The EM algorithm sets itself apart from other commonly used methods by taking
into account all possible values taken by the latent variable $S$. To do
so, it computes the expected value of the log likelihood of $\theta$,
conditioned by the posterior distribution
$\mathbb{P}_{\hat{\theta}_{q-1}} (S|X)$, also named as the *auxiliary
function*. Utilising the assumption of independence among observations in a mixture model, the general formula of this proxy
function of the incomplete log-likelihood is given in finite mixture
models by Equation \@ref(eq:8).

```{=tex}
\begin{equation}
\begin{split}
Q(\theta|\hat{\theta}_{q-1}) & := \mathbb{E}_{S_{1:n}| X_{1:n}, \hat{\theta}_{q-1}} \left[\ell(\theta | X_{1:n}, S_{1:n})\right] \\
&=\sum_{i=1}^n \sum_{j=1}^k \eta_{i}(j) \left( \log (p_j) +  \log (\mathbb{P}(X_i|S_i=j, \theta)) \right)
\end{split}
(\#eq:8)
\end{equation}
```
with $\hat{\theta}_{q-1}=\hat{\theta}$ the current estimated parameter
value.

In practice, the EM algorithm consists in performing alternatively E-step and M-step until convergence, as described in the pseudocode below (Box 1):

::: {.blackbox data-latex="{\\textbf{Box 1: the EM algorithm}}"}
::: {.center data-latex=""}
-   *step E*: determine the posterior probability function $\eta_i(j)$
    for each observation of $X$ for each possible discrete latent class,
    using the initial estimates $\hat{\theta}_0$ at step $q=0$, or the
    previously computed estimates $\hat{\theta}_{q-1}$. The general formula is given by Equation \@ref(eq:10):

```{=tex}
\begin{equation}
    \eta_i(j) = \frac{p_j f_{\zeta_j} (x_i)}{\sum_{j=1}^k p_j f_{\zeta_j} (x_i)}
(\#eq:10)
\end{equation}
```

-   *step M*: compute the mapping function
    $\hat{\theta}_q:=M(\theta | \hat{\theta}_{q-1})=\arg \max Q(\theta| \hat{\theta}_{q-1})$ which maximises the auxiliary function. One way of retrieving the MLE associated to the auxiliary function is to determine the roots of its derivative, namely solving Equation \@ref(eq:mapping-function-derivative)^[To ensure
that we reach a maximum, we should assert that the Hessian matrix evaluated at the MLE is indeed negative definite.]:
```{=tex}
\begin{equation}
    \frac{\partial Q(\theta| \hat{\theta}_{q-1})}{\partial \theta}=0
(\#eq:mapping-function-derivative)
\end{equation}
```

:::
:::

Interestingly, the decomposition of the incomplete log-likelihood
associated to a mixture model $\ell(\theta|X)$ reveals an entropy term
and the so-called auxiliary function [@dempster_etal77]. It can be used to prove that
maximising the auxiliary function at each step induces a bounded
increase of the incomplete log-likelihood. Namely, the convergence of
the EM algorithm, defined by comparisons of consecutive log-likelihood,
is guaranteed, provided the mapping function returns the maximum of the
auxiliary function. Yet, the convergence of the series of estimated
parameters
$(\theta_q)_{q \ge 0} \underset{i\to +\infty}{\longrightarrow} \hat{\theta}$
is harder to prove but has been formally demonstrated for the *exponential family* (a superset of the Gaussian family), as stated in @dempster_etal77.

Additionally, the EM algorithm is *deterministic*, meaning that for a
given initial estimate $\theta_0$ the parameters returned by the
algorithm at a given step $q$ are fixed. However, this method requires the user to provide an initial estimate, denoted as $\theta_0$, of the model parameters and to specify the number of components in the mixture. We review some classic
initialisation methods in [Initialisation of the EM algorithm] and some
algorithms used to overcome the main limitations of the EM
algorithm in the Appendix *Extensions of the EM algorithm to overcome its limitations*.

Finally, the prevalent choice of Gaussian distributions to characterize the distribution of random observations is guided by a set of interesting properties. In particular, @geary36 has shown that the Normal distribution is the only distribution for which the Cochran's theorem [@cochran34] is guaranteed, namely for which the the mean and variance of the sample are independent of each other. Additionally, similar to any distribution proceeding from the exponential family, the MLE statistic is *sufficient*^[The Pitman–Koopman–Darmois theorem [@koopman36] states that only the exponential family provides distributions whose statistic can summarize arbitrary amounts of iid draws using a finite number of values].



## Initialisation of the EM algorithm

<!-- Precisely, the rebmix algorithm initiates the estimation of another component when at least $25\%$ of the unassigned intervals are assigned to the currently estimated component. -->

EM-like algorithms require an initial estimate of the parameters, $\theta_0$, to
optimise the maximum likelihood. *Initialisation* is a crucial step, as
a bad initialisation can possibly lead to a local sub-optimal solution
or trap the algorithm in the boundary of the parameter space. The most straightforward initialisation methods, such as random initialisation, are standalone and do not require any additional initialisation algorithms, whereas *meta-methods*, such as short-EM, still need to be initialised by alternative methods. The commonly-used initialisation methods encompass:

-   The *Model-based Hierarchical Agglomerative Clustering* (MBHC) is
    an agglomerative hierarchical clustering based on MLE criteria applied to
    GMMs [@scrucca_raftery15]. First, the MBHC is
    initialised by assigning each observation to its own cluster. Next,
    the pair of clusters that maximises the likelihood of the underlying
    statistical model among all possible pairs is merged. This procedure
    is repeated until all clusters are merged. The final resulting
    clusters are then simply the last $k$ cuts of the resulting
    dendrogram. When the data is univariate and homoscedastic, or when the underlying distribution has a diagonal covariance matrix, the merging criterion performs similarly to *Ward's criterion*, in that merging of the two clusters also simultaneously minimizes the sum of squares.
    As opposed to the other initialisation methods described hereafter, MBHC is a
    deterministic method which does not require careful calibration of
    hyperparameters. However, as acknowledged by the author of the method [@fraley98], the resulting partitions are generally suboptimal compared to other initialisation methods.

-   The conventional *random* initialization method, frequently employed for the initialization step of the *k*-means algorithm, involves the random selection of $k$ distinct observations, which are referred to as *centroids*. Subsequently, each observation is assigned to the nearest centroid, a process reminiscent of the C-step in the CEM algorithm [@biernacki_etal03]. This is the method used in this paper, unless otherwise stated.  Alternative versions of this method have been developed: for instance, the package \pkg{mixtools} draws the proportions of the components from a Dirichlet distribution, whose main advantage lies in respecting the unit simplex constraint (Equation \@ref(eq:2))^[Without prior knowledge favouring one component over another, the Dirichlet distribution is generally parametrised by $\alpha=\frac{1}{k}$, implicitly stating that any observation has equal chance to proceed from a given cluster. In that case, the corresponding distribution is parametrised by a single scalar value $\alpha$, called the *concentration parameter*.], but uses binning methods to guess the means and standard deviations of the components. Similarly, @kwedlo13 proposes a method in which the means of the components are randomly chosen, but with an additional constraint of maximising the Mahalanobis distance between the selected centroids. This enables to cover a larger portion of the parameters' space.

-   *k*-means is a CEM algorithm, in which the additional assumption of
    balanced classes and homoscedascity implies that each observation in
    the E-step is assigned to the cluster with the nearest mean (the one
    with the shortest Euclidean distance). *K*-means is initialised by randomly selecting $k$ points, known as the *centroids*. It is often chosen for its fast convergence and memory-saving consumption.

-   The *quantile* method sorts each observation $x_i$ in an increasing
    order and splits them into equi-balanced quantiles of size $1/k$. Then, all
    observations for a given quantile are assumed to belong to the same
    component. \footnote{This method is only available in the univariate framework,
    since it is not possible to define a unique partition of the observable space into $k$-splits. For example, in bivariate setting, a binning with $k=2$ components on each axis leads to a total of $2 \times 2=4$ binned regions, which raises the selection issue of the best $k$ hyper-squared volumes for the initial parameters estimation. More generally, $\binom{D}{k}$ binning choices are possible in the multivariate setting.}

-   The *Rough-Enhanced-Bayes mixture* (REBMIX) algorithm is implemented in the \CRANpkg{rebmix} [@R-rebmix] package and the complete
    pseudo-code is described thoroughly in [@rebmix2015a; @rebmix2020b]. The key stages implemented by the rebmix algorithm for initialising the parameters of GMMs encompass:
    + First, the observations are processed using one of these three methods: *k*-nearest neighbours (KNN), Parzen kernel density estimation, or binned intervals. With the binned interval method, the observations are initially divided into $\sqrt{n}^D$ intervals of equal lengths. The mode of one of the components' distribution is subsequently determined by the midpoint of the interval with the highest frequency. The observations lying within the interval are used as preliminary estimates, referred to as "rough" parameters in @rebmix2015a.
    + All other observations and intervals are then iteratively assigned to the currently estimated component or to residual components, the ones that have not yet been characterised. The decision to assign an interval to either the currently estimated component or one of the residual components depends on the magnitude of the discrepancy between the observed and the expected frequency within the interval.
    + Finally, all intervals assigned to the currently estimated component (and not only the interval including the mode of the distribution) are used to determine the parameters of the associated Gaussian distribution. Since this step relies on a more comprehensive number of observations for parameter estimation, guaranteeing in principle more robust estimates, this stage is referred to as "enhanced" estimation in @rebmix2015a. The algorithm terminates when all intervals have been assigned to a cluster, and the parameters of the various distribution components have been estimated.
    
The rebmix algorithm can thus be seen as a natural extension of the quantile method, with more rigorous statistical support. Two drawbacks of the algorithm include the need for intensive calibration of hyperparameters and its inadequacy for the estimation of highly overlapping or high dimensional mixture distributions^[The method we describe here to preprocess the observations in order to estimate the empirical density estimation, namely the "histogram method" is not well suited for high dimensional data, as the exponential growth of the volume with respect to dimensionality leads to data sparsity, related to the well-known issue of the "curse of dimensionality". Indeed, $\sqrt{n}^D$ distinct intervals will be parsed by the method and the probability with an increasing number of features and decreasing number of observations that no clear local maximum emerges converges to 1. In high-dimensional context, the Parzen window or the KNN method should be favoured, see [@rebmix2015a], p. 16.].

-   The *meta-methods* consist generally in short runs of EM-like
    algorithms, namely CEM, SEM and EM (see Appendix B: *Extensions of the EM algorithm to overcome its limitation*), with alleviated convergence criterion. The main idea is
    to use several random initial estimates with shorter runs of the
    algorithm to explore larger regions of the parameter space and avoid being trapped in a local maximum. Yet, these methods are highly dependent on the choice of the initialisation algorithm [@biernacki_etal03].

-   In the high-dimensional setting, if the number of dimensions $D$ exceeds the number of observations $n$, all previous methods must be adjusted, usually by first projecting the dataset into a smaller, suitable subspace and then inferring prior parameters in it. In particular, \pkg{EMMIXmfa}, in the mixture of common factor analysers (MCFA) approach, initialises the shared projection matrix $\boldsymbol{Q}$ by either keeping the first $d$ eigen vectors generated from standard principal component analysis or uses custom random initialisations [@baek_etal10].

Following this theoretical introduction, we empirically evaluate the performance of the aforementioned R packages, considering various initialization algorithms and the complexity of the GMMs distributions.
Precisely, we outline the simulation framework used to compare the seven packages in [Methods] and report the results in [Results]. We conclude by providing a general simplified framework to select the combination of package and initialisation method best suited to its objectives and the nature of the distribution of the dataset.

# A comprehensive benchmark comparing estimation performance of GMMs

We searched CRAN and Bioconductor mirrors for packages that can retrieve
parameters of GMM models. Briefly, out of 54 packages dealing with GMMs
estimation, we focused on seven packages that all estimate the
MLE in  GMMs using the EM algorithm, were recently
updated and allow the users to specify their own initial estimates:
\pkg{bgmm}, \pkg{EMCluster}, \pkg{flexmix},
\pkg{GMKMcharlie}, \pkg{mclust}, \pkg{mixtools} and \pkg{Rmixmod}. The
complete inclusion process is detailed in Appendix C, *the meta-analysis workflow for the final selection of CRAN and Bioconductor platforms*. The flowchart summarising our choices is represented in Figure \@ref(fig:flowchart).

```{r flowchart, fig.cap="A minimal roadmap used for the selection of the packages reviewed in our benchmark.", fig.alt="From root to top, we describe schematically the filtering process used for the final selection of the reviewed packages"}
knitr::include_graphics("figures/flowchart_packages_selection.png")
```


```{r retrieve-general-information, echo=FALSE, eval=FALSE}
relevant_packages <- c("EMMIXmfa", "HDclassif")
info_package <- relevant_packages %>%
  purrr::map_dfr(function(x) {
    description <- utils::packageDescription(
      pkg = x,
      fields = c(
        "Package", "Version", "Date", "Maintainer",
        "License", "Imports", "Depends", "LinkingTo"
      )
    )
    description_tibble <- tibble::tibble(
      Package = description$Package, Version = description$Version, Date = description$Date,
      Maintainer = description$Maintainer, License = description$License, Imports = description$Imports,
      Depends = description$Depends, LinkingTo = description$LinkingTo
    )
    return(description_tibble)
  })

info_package <- cranlogs::cran_downloads(
  package = relevant_packages,
  from = "2019-04-01", to = "2023-04-01"
) %>%
  dplyr::group_by(package) %>%
  dplyr::summarise(Popularity = round(mean(count))) %>%
  dplyr::inner_join(info_package, by = c("package" = "Package"))

info_package <- info_package %>% dplyr::inner_join(
  purrr::imap_dfr(
    tools::package_dependencies(packages = relevant_packages, recursive = TRUE),
    ~ tibble::tibble(package = .y, recursive_deps = list(.x))
  )
)
```



We also include two additional packages dedicated specifically to high-dimensional settings, namely \CRANpkg{EMMIXmfa} [@R-EMMIXmfa] and \CRANpkg{HDclassif} [@R-HDclassif] to compare their performance with standard multivariate approaches in complex, but non degenerate cases. We summarise the main features and use cases of the seven + two reviewed
packages in Table
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:low-comparison-package-html)', '\\@ref(tab:low-comparison-package-pdf)'))`.
The three most commonly used packages are \pkg{mixtools}, \pkg{mclust} and \pkg{flexmix}. However, the \pkg{mclust} package is by far the most complete with many features provided to visualise and evaluate the quality of the GMM estimate. \pkg{bgmm} has the greatest number of dependencies, while \pkg{mclust} only depends of base R packages. Additionally, in parallel to clustering tasks, \pkg{flexmix} and \pkg{mixtools} packages perform
regression of mixtures and implement mixture models using other
parametric distributions or non-parametric methods via kernel-density
estimation.



```{r low-comparison-package-html, layout="l-body-outset", eval = knitr::is_html_output()}
readr::read_delim("./tables/package_comparison_low_level.csv", delim = ";", show_col_types = F) %>%
  mutate(
    Imports = stringr::str_replace(Imports, ">=", "\U2265"),
    dplyr::across(.cols = c("Regression", "Implemented models"), ~ gsub(pattern = "redcross", "./figures/red_cross.png", .x)),
    dplyr::across(.cols = c("Regression", "Implemented models"), ~ gsub(pattern = "greentick", "./figures/green_tick.png", .x)),
    across(.cols = where(is.character), stringr::str_replace_all, "\\\\n", " ")
  ) %>%
  flextable() %>%
  flextable::mk_par(i = 1:9, j = 3, value = as_paragraph(
    flextable::as_image(src = Regression, width = 0.5, height = 0.5, unit = "cm")
  )) %>%
  flextable::mk_par(i = c(1, 4:9), j = 4, value = as_paragraph(
    flextable::as_image(src = "./figures/red_cross.png", width = 0.5, height = 0.5, unit = "cm")
  )) %>%
  set_caption(caption = "Main features of the reviewed packages, sorted by decreasing number
of daily downloads. *Downloads per day* returns the daily average number of downloads for each package on the last 2 years.
 *Recursive dependencies* column counts the complete set of non-base packages required, as first-order dependencies depend on other packages as well.", html_escape = F) %>%
  theme_vanilla() %>%
  align(align = "center", part = "header") %>%
  flextable::autofit() %>%
  flextable::hline(i = 7)
```

```{r low-comparison-package-pdf, eval = knitr::is_latex_output()}
readr::read_delim("./tables/package_comparison_low_level.csv", delim = ";", show_col_types = F) %>%
  mutate(dplyr::across(.cols = c("Regression", "Implemented models"), ~ gsub(pattern = "redcross", "$\\\\redcross$", .x)),
    dplyr::across(.cols = c("Regression", "Implemented models"), ~ gsub(pattern = "greentick", "\\\\greentick", .x)),
    Imports = gsub(pattern = ">=", "$\\\\ge$", Imports),
    across(.cols = where(is.character), kableExtra::linebreak, linebreaker = "\\\\n")
  ) %>%
  kbl(
    booktabs = T, caption = "Main features of the reviewed packages, sorted by decreasing number of daily downloads.
               \\textit{Downloads per day} returns the daily average number of downloads for each package on the last 2 years.
               \\textit{Recursive dependencies} column counts the complete set of non-base packages required,
               as first-order dependencies depend on other packages as well.", escape = F,
    col.names = linebreak(c(
      "Package", "Version",
      "Regression", "Implemented \n models", "Downloads \n per day",
      "Last \n update", "Imports", "Recursive \n dependencies", "Language"
    ))
  ) %>%
  row_spec(0, bold = T, align = "c", hline_after = T) %>%
  row_spec(7, hline_after = T) %>%
  kable_styling(latex_options = c("hold_position", "scale_down"))
```

We further detail features specifically related to GMMs in Table
`r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:high-comparison-packages-html)', '\\@ref(tab:high-comparison-packages-pdf)'))`. We detail row after row its content below:

- The parametrisations used to provide parsimonious estimation of the GMMs are reviewed in [Parameter estimation in finite mixtures models] and summarised in rows 1 and 2 (Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:high-comparison-packages-html)', '\\@ref(tab:high-comparison-packages-pdf)'))`) for the univariate and multivariate setting. We refer to the package as "canonical" when it implements both homoscedastic and heteroscedastic parametrisations in the univariate setting, and the 14 parametrisations listed in Supplementary Table 3 in the multivariate setting. Adding the additional constraint of equi-balanced clusters results in a total to $14 \times 2=28$ distinct models and $2 \times 2=4$ parametrisations, respectively in the univariate and multivariate setting. Since \pkg{EMMIXmfa} and \pkg{HDclassif} are dedicated to the analysis of high-dimensional datasets, they project the observations in a smaller subspace and are not available in the univariate setting. Given an user-defined or prior computed intrinsic dimension, we can imagine using any of the standard parametrisations available for instance in the \pkg{mclust} package, and listed in Appendix *Parsimonious parametrisation of multivariate GMMs*. In addition, \pkg{HDclassif} allows each cluster $j$ to be represented with its own subspace intrinsic dimension $d_j$, as we describe in further details in Appendix *Parameters estimation in a high-dimensional context*.

- [The EM algorithm] is the most commonly employed method for estimating the parameters of GMMs, however, alternative algorithms based on the EM framework, are reviewed in Appendix B: *Extensions of the EM algorithm to overcome its limitations* and row 3 of Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:high-comparison-packages-html)', '\\@ref(tab:high-comparison-packages-pdf)'))`. Especially, GMMs estimation is particularly impacted by the presence of outliers, justifying a specific benchmark (see Appendix *A small simulation to evaluate the impact of outliers*).
We briefly review the most common initialisation algorithms in section [Initialisation of the EM algorithm] and row 4 of Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:high-comparison-packages-html)', '\\@ref(tab:high-comparison-packages-pdf)'))`, a necessary and tedious task for both the EM algorithm and its alternatives.

- To select the best parametrisations and number of components that fit the mixture, several metrics are provided by the reviewed packages (*Model selection* and row 5). Due to the complexity of computing the true distribution of the estimated parameters, bootstrap methods are commonly used used to derive confidence intervals (see Appendix *Derivation of confidence intervals in GMMs* and row 6 in Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:high-comparison-packages-html)', '\\@ref(tab:high-comparison-packages-pdf)'))`).

<!-- - To generate random samples and build bootstrap confidence intervals of the estimated parameters, we first need to design a function that mimics the effect of drawing a random observation from a GMM, given user defined parameters. Packages able to do so are gathered as parametric. An alternative way consists of providing the user either real or toy datasets, the last ones having the major advantage that we know the expected values of the parameters. -->
<!-- However, especially in high-dimensional settings, it is rather cumbersome to provide each of the parameters required to uniquely identify a GMM. Indeed, without any constraint on the parameters and full covariance matrices, a multivariate GMM of dimension $D$ and with $k$ components requires $k - 1 +kD +\frac{D(D+1)}{2}$ free parameters. \pkg{bgmm} is the only package among the ones we benchmarked that partly : in the absence of prior information from the user, the proportions are supposed to be all equal, the mean values are drawn independently from an univariate Normal distribution and the covariance matrix is set equal to the sample covariance of a $D$ independent-wise Normal draw^[Expecting the user to provide their own estimates is relatively expected in packages that only estimate parameters using variants of maximum likelihood techniques. Indeed, in such methods, no prior distribution of the parameters is assumed, in opposition to Bayesian approaches].  -->

- Six packages supply several functions for visualisation, summarised in the last row of Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:high-comparison-packages-html)', '\\@ref(tab:high-comparison-packages-pdf)'))`, to display either the distributions corresponding to the estimated parameters or compare quickly the performance across packages. However, \pkg{mclust} is by far the most complete one, with density plots (in the univariate setting) and isodensity plots (bi-dimensional in the bivariate setting or in higher dimensions after appropriate dimensionality reduction), with the option to plot custom confidence intervals and critical regions, and finally boxplot bootstrap representations for displaying the distribution of the benchmarked estimated parameters.

High-dimensional packages provide specific representations adjusted to the high-dimensional settings, notably allowing the user to visualise the projected factorial representation of its dataset in a two or three-dimensional subspace. They also provide specialised performance plots, notably scree plots or BIC scatter plots to represent in a compact way numerous projections and parametrisations.


```{r high-comparison-packages-html, layout="l-body-outset", eval = knitr::is_html_output()}
data <- readr::read_delim("./tables/package_comparison_high_level.csv", delim = ";", show_col_types = F) %>%
  filter(!(Column1 %in% c("Labelled data", "Weighted implementation", "Simulation"))) %>%
  tibble::column_to_rownames("Column1") %>%
  mutate(
    dplyr::across(.cols = everything(), ~ gsub(pattern = "redcross", "./figures/red_cross.png", .x)),
    dplyr::across(.cols = everything(), ~ gsub(pattern = "greentick", "./figures/green_tick.png", .x))
  ) %>%
  tibble::rownames_to_column("Package features")

flex_data <- data %>% flextable()
for (i in 1:nrow(data)) {
  green_index <- grep("green_tick", data[i, ] %>% unlist())
  red_index <- grep("red_cross", data[i, ] %>% unlist())
  flex_data <- flex_data %>%
    flextable::mk_par(
      i = i, j = green_index,
      value = as_paragraph(flextable::as_image(
        src = "./figures/green_tick.png",
        width = 0.5, height = 0.5, unit = "cm"
      ))
    ) %>%
    flextable::mk_par(
      i = i, j = red_index,
      value = as_paragraph(flextable::as_image(
        src = "./figures/red_cross.png",
        width = 0.5, height = 0.5, unit = "cm"
      ))
    )
}

flex_data %>%
  set_caption(
    caption = "Custom features associated to GMMs estimation for any of the benchmarked packages.",
    html_escape = F
  ) %>%
  theme_vanilla() %>%
  align(align = "center", part = "header") %>% # centre align header
  flextable::autofit() %>%
  flextable::vline(j = 8, border = officer::fp_border(width = 2))
```

```{r high-comparison-packages-pdf, eval = knitr::is_latex_output()}
readr::read_delim("./tables/package_comparison_high_level.csv", delim = ";", show_col_types = F) %>%
  filter(!(Column1 %in% c("Labelled data", "Weighted implementation", "Simulation"))) %>%
  tibble::column_to_rownames("Column1") %>%
  mutate(
    dplyr::across(.cols = everything(), ~ gsub(pattern = "redcross", "\\\\redcross", .x)),
    dplyr::across(.cols = everything(), ~ gsub(pattern = "greentick", "\\\\greentick", .x))
  ) %>%
  kbl(booktabs = T, caption = "Custom features associated to GMMs estimation for any of the benchmarked packages.", escape = F) %>%
  kable_styling(latex_options = c("hold_position", "scale_down")) %>%
  row_spec(0, bold = T, align = "c") %>%
  row_spec(1:7, hline_after = T) %>%
  column_spec(column = 1:10, latex_valign = "m", width = c("1.6cm", "4cm", rep("1.6cm", 9))) %>%
  column_spec(column = 1, bold = T, border_right = T) %>%
  column_spec(column = 8, border_right = T)
```


## Methods

In addition to the the seven packages selected for our benchmark, we
include a custom R implementation of the EM algorithm used
as baseline, referred to as *RGMMBench*, and for the high-dimensional setting we select packages \pkg{EMMIXmfa} and \pkg{HDclassif}, on the basis of criteria detailed in Appendix C, *General workflow*. Code for *RGMMBench* is provided in Appendix *Application of the EM algorithm to GMMs*. To compare the statistical performances of these packages, we performed *parametric bootstrap* (*Derivation of confidence intervals in GMMs*) and built an experimental design to cover distinct mixture distributions parameter configurations, using prior user-defined parameters.


For each experiment, we assign each observation to an unique cluster by drawing $n$ labels $S_{1:n}$ from a multinomial distribution whose parameters were the prior user-defined proportions $p=(p_1, \ldots, p_k)$. Then, each observation $x_i$ assigned to hidden component $j$ is drawn from a Normal distribution using the `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("stats::rnorm()"),   '\\code{stats::rnorm()}'))` function for the univariate distribution and `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("MASS::mvrnorm"),   '\\code{MASS::mvrnorm}'))` for the multivariate distribution. The complete code used for simulating data is
available on GitHub at [RGMMBench](https://github.com/bastienchassagnol-servier/RGMMBench). Finally, we obtain an empirical distribution of the estimated parameters by computing the MLE
of each randomly generated sample.

For all the packages, we used the same convergence threshold, $10^{-6}$, and maximum of 1,000 iterations, as a numerical criterion for convergence. We generated simulated data with $n = 200$ observations in the univariate setting and $n = 500$ observations in the bivariate setting. We set the number of observations in order to minimise the probability of generating a sample without drawing any observations from one of the components^[It is especially critical in cases of highly unbalanced configurations, as detailed in Appendix *Practical details for the implementation of our benchmark*]. Unless stated explicitly, we kept the default hyper-parameters and custom global options provided by each package. For instance, the \pkg{flexmix} package has a default option, *minprior*, set by default to 0.05, which removes any component present in the mixture with a ratio below $0.05$. Besides, the fully unconstrained model was the only one which we implemented both in the univariate and multivariate settings, as it is the only parametrisation implemented in all the seven packages.


We compared the packages' performances using five
initialisation methods: random, quantile, *k*-means, rebmix and hierarchical clustering in the univariate setting. We benchmarked the same initialisation methods in the multivariate setting, except for the quantile method which has no multivariate equivalent (see section [Initialisation of the EM algorithm]):


  * We used the function `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("EMCluster::rand.EM()"),   '\\code{EMCluster::rand.EM()}'))` with 10 random restarts and minimal cluster size of 2 for the random initialisation. The method implemented by \pkg{EMCluster} is the most commonly used, described in details in @biernacki_etal03 and in section [Initialisation of the EM algorithm].

  * To implement the k-means initialisation, we used the `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("stats::kmeans()"),   '\\code{stats::kmeans()}'))` function with a convergence criterion fo 10^{-2} and maximum of 200 iterations. The initial centroid and covariance matrix for each component were computed by restricting to the sample observations assigned to the corresponding component. The approach is close to the one adopted by the CEM algorithm (see Appendix B: *Extensions of the EM algorithm to overcome its limitations*).

  * We used the `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("mclust::hcV()"),   "\\code{mclust::hcV()}"))` function for the MBHC algorithm. This method has two main limitations: just like the *k*-means implementation, it only returns a cluster assignment to each observation instead of the posterior probabilities, and the splitting process to generate the clusters sometimes results in clusters composed of only one observation. To avoid this, we added a small epsilon to each posterior probability.

  * We used in the univariate setting `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("bgmm::init.model.params"),   '\\code{bgmm::init.model.params}'))` for the quantiles initialisation.

  * To implement the rebmix method, we used the `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("rebmix::REBMIX()"),   '\\code{rebmix::REBMIX}'))` function, using the *kernel density estimation* for the estimation of the empirical density distribution coupled with *EMcontrol* set to one to prevent the algorithm from starting EM iterations.

  * Any of the seven packages could be used to implement the small EM method. We decided to use the `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("mixtools::normalmixEM"),   '\\code{mixtools::normalmixEM}'))` as it is the closest one to our custom implementation. We specified 10 random restarts, a maximal number of iterations of 200 and an alleviated absolute threshold of $10^{-2}$. Preliminary experiments have led us to consider the removal of the small EM initialization method from the simulation benchmark. This decision is based on the observation that the differences of performance observed between the packages were no longer significant (see supplementary Figure 9).


We sum up in Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:general-parameter-description-html)', '\\@ref(tab:general-parameter-description-pdf)'))` the general configuration used to run the scripts. Additionally, all simulations were run with the same R [@R-core] version 4.0.2 (2020-06-22). 

<!-- using an OS system under Linux and numerical precision of $2.22 \times 10^{-16}$. -->


```{r general-parameter-description-html, eval=knitr::is_html_output()}
read.table("./tables/general_simulation_parameters.csv", sep = ";", header = T, check.names = F) %>%
  flextable() %>%
  compose(j = "Criterion threshold", value = as_paragraph(as_equation(`Criterion threshold`, width = 2, height = .5))) %>%
  set_caption(caption = "Global options shared by all the benchmarked packages.", html_escape = F) %>%
  theme_vanilla() %>%
  align(align = "center", part = "header") %>%
  flextable::autofit()
```

```{r general-parameter-description-pdf, eval=knitr::is_latex_output()}
formatted_data <- readr::read_delim("./tables/general_simulation_parameters.csv", delim = ";") %>%
  dplyr::mutate(`Criterion threshold` = paste0("$", `Criterion threshold`, "$"))


formatted_data %>%
  kbl(booktabs = T, caption = "Global options shared by all the benchmarked packages.", escape = F, midrule = "midrule", align = "c") %>%
  kable_styling(latex_options = c("hold_position", "scale_down")) %>%
  row_spec(0, bold = T) %>%
  row_spec(1, hline_after = T) %>%
  column_spec(1:ncol(formatted_data), width = "4.1cm") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```


Preliminary experiments suggested that the quality of the estimation of a GMM is mostly affected by the overlap between components' distribution and level of unbalance between components. We quantified the overlap between two components by the following overlap score (OVL, see Equation \@ref(eq:overlap)), with a smaller score denoting well-separated components:

```{=tex}
\begin{equation}
       \operatorname{OVL}(i, j) = \int \min (f_{\zeta_i} (x), f_{\zeta_j} (x)) dx \quad \text{ with } i \neq j
  (\#eq:overlap)
\end{equation}
```
We may generalise this definition to $k$ components by averaging the
individual components' overlap. We use the function
`r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("MixSim::overlap"),   '\\code{MixSim::overlap}'))` from the \CRANpkg{MixSim} package [@R-MixSim] that approximates this quantity using a Monte-Carlo based method (see appendices *An analytic formula of the overlap for univariate Gaussian mixtures* and *Practical details for the implementation of our benchmark* for further details).

The level of imbalance may be evaluated with entropy measure (Equation \@ref(eq:entropy)):


```{=tex}
\begin{equation}
    H(S)=-\sum_{j=1}^k p_j \log_k (p_j)
(\#eq:entropy)
\end{equation}
```
with $k$ is the number of components and $p_j=\mathbb{P}(S=j)$ is the frequency of class $j$.


We considered 9 distinct configuration parameters, associated with distinct values of $\operatorname{OVL}$ and entropy in the univariate setting, 20 configurations in the bivariate setting, and 16 configurations in the high-dimensional setting. Briefly, in the univariate setting, we simulated components with the same set of four means (0, 4, 8, and 12), three sets of mixture proportions \
$\left[(0.25, 0.25, 0.25, 0.25); (0.2, 0.4, 0.2, 0.2); (0.1, 0.7, 0.1, 0.1) \right]$ and three variances: $(0.3, 1, 2)$. In the bivariate setting, we consider two sets of proportions: $\left[(0.5, 0.5); (0.9, 0.1) \right]$, two sets of coordinate centroids: $\left[(0; 20), (20, 0) \right]$ and $\left[(0; 2), (2, 0) \right]$, the same variance of 1 for each feature and for each component for illustrative purposes of the direct relation linking the correlation and the level of OVL and five sets of correlation:$[(-0.8, -0.8); (0.8, -0.8); (-0.8, 0.8); (0.8, 0.8); (0, 0)]$.

Finally, we tested eight configurations in the high-dimensional framework, setting to $D=10$ the number of dimensions. We modified the level of overlap (definition is reported in Equation \@ref(eq:overlap)) and the imbalance between the component proportions across our simulations. Additionally, we tested two types of constraints on the covariance matrix: fully parametrised and spherical (see Appendix *Parsimonious parametrisation of multivariate GMMs*). Each of the parameter configurations tested in the high-dimensional setting was evaluated with $n=200$ observations and $n=2000$ observations. Additionally, instead of manually defining the parameters for the high-dimensional simulation, we used the `r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("MixSim::MixSim"),   '\\code{MixSim}'))` function from the \CRANpkg{MixSim} package [@R-MixSim]. This function returns the user a fully parametrised GMM, with a prior defined level of maximum or average overlap^[Unfortunately, as discussed in further details in Appendix *An analytic formula of the overlap for univariate Gaussian mixtures*, the MixSim package does not compute the global distribution overlap, but instead returns the mean of pairwise overlap between any component (however, with two components, these two alternative definitions match.) Finally, it is not possible to set the proportions of the components before the generation of the parameters, except for clusters with equal proportions, and contrary to the expect behaviour of additional parameter *PiLow*, supposed to define the smallest mixing proportion.].


The complete list of parameters used is reported respectively in Table \@ref(tab:parameter-configuration-univariate) for the univariate setting, Table \@ref(tab:parameter-configuration-bivariate) for the bivariate setting and \@ref(tab:parameter-configuration-HD) for the high-dimensional setting. We benchmarked simulations where the components were alternatively very distinct or instead very overlapping, as well as of equal proportions or instead very unbalanced. The adjustments made to meet the specific requirements of high dimensional packages are detailed in *Practical details for the implementation of our benchmark*.


We report the most significant results and features and the associated recommendations in next section [Results].

## Results

All figures and performance overview tables are reported in *Supplementary Figures and Tables in the univariate simulation* for the univariate setting, *Supplementary Figures and Tables in the bivariate simulation* for the bivariate scenario and *Supplementary Figures and Tables in the HD simulation* for the high dimensional scenario.

**Balanced and non-overlapping components**


In the univariate setting, with balanced components and low $\operatorname{OVL}$ (scenario U1 in Table
\@ref(tab:parameter-configuration-univariate)), the parameter
estimates are identical in most cases across initialisation methods and packages, notably the same estimates are returned with *k*-means or rebmix initialisation. However, the random initialisation method leads
to a higher variance and bias on the parameter estimates than other
methods (Supplementary Figure 4 and Supplementary Table 6),
with some estimates fitting only local maxima, far from the optimal value.


Similarly, the scenarios in the bivariate setting  (configurations B6-B10 in Table \@ref(tab:parameter-configuration-bivariate)), with a focus on B6, B7 and B10 visualised in Supplementary Figure 16, feature well-separated and balanced components. Consistent with conclusions from the corresponding univariate configurations, all benchmarked packages return the same estimates across initialisation methods.


**Unbalanced and non-overlapping components**


However, with unbalanced classes and low $\operatorname{OVL}$ (scenario U7 in
\@ref(tab:parameter-configuration-univariate)), the choice of the
initialisation method is crucial, with quantiles and random
methods yielding more biased estimates and proned to fall in local maximum. Rebmix initialisation provides the best estimates, with the smallest MSE and bias across packages (Supplementary Figure
5 and supplementary Table 7, always associated with the highest likelihood. Overall, with
well-discriminated components, most of the differences on the estimation
originate from the choice of initialisation method, while the choice of
the package has only small impact.

In the bivariate framework, two configurations featured both strongly unbalanced and well-separated components, similarly to scenario U3 in Table \@ref(tab:parameter-configuration-univariate): the configurations B12 (Supplementary Figure 12 and Table 12) and B14 (Supplementary Figure 13 and Supplementary Table 13). Similarly, configurations B16, B17 and B20 (Table \@ref(tab:parameter-configuration-bivariate)) with similar characteristics are summarised in supplementary Figure 17. In all these configurations, neither the initialisation method nor the package have a statistical significant impact on the overall performance.


Similarly, configurations HD1a-HD4b in Table \@ref(tab:parameter-configuration-HD)) in the high dimensional setting display well-separated clusters, with a representative outcome represented in Supplementary Figure 19 and Supplementary Table 16. Consistent with the results obtained in the analogous univariate and bivariate scenarios, in the unbalanced and non-overlapping framework, the majority of the benchmarked packages produce highly consistent and similar estimates when hierarchical clustering and k-means were used for parameter initialisation. However, \pkg{bgmm} and \pkg{EMCluster} clearly perform worse when the rebmix initialization method is used (however, overall, rebmix performs poorly, regardless of the package used for estimation). Notably, initialisations with the rebmix package tend to display a much larger number of poor estimations, some of which can be identified with the local maxima associated with parameter switching between the two classes. Finally, the two additional packages dedicated to high-dimensional clustering display the worst performances, with \pkg{EMMIXmfa} returning the most biased parameters and \pkg{HDclassif} the most noisy estimates. \pkg{EMMIXmfa} is the only package that returned highly biased estimates of the components' proportions in this setting.



**Balanced and overlapping components**


When the overlap between components increases, the bias and variability of the estimates tends to increase, and the choice of initialisation method becomes more impactful.
The least biased and noisy estimations with balanced components in the univariate setting (scenario U3 in Table
\@ref(tab:parameter-configuration-univariate)) are obtained with
the *k*-means initialisation (Supplementary Figure
3 and Table 8) while the rebmix initialisation returns the most biased and noisy
estimates. Similar results are found in the bivariate setting with a balanced and highly overlapping two-component GMM (configurations B1-B5 from Table \@ref(tab:parameter-configuration-bivariate)), with the best performance reached with the *k*-means initialisation method, followed by MBHC. This is emphasised in supplementary Figure 16, in the top three most complex configurations, namely B1, B2 and B5. If the shape of the covariance matrix is well-recovered, no matter the package, the Hellinger distances are significantly higher (and thus the estimate further away from the target distribution) with the random and rebmix methods.



Similarly, in the high-dimensional scenario HD7 of Table \@ref(tab:parameter-configuration-HD)), presenting balanced but highly overlapping clusters with a full covariance structure, the best performance was obtained with *k*-means initialisation, while the rebmix initialisation returned the most biased and noisy estimates. While \pkg{EMMIXmfa} performed well when it converged, it returned an error in most cases (see Column *Success* of supplementary Table 17). The least biased estimates were returned by \pkg{mixtools} and \pkg{Rmixmod} and the least noisy by \pkg{flexmix}, \pkg{mclust} and \pkg{GMKMCharlie} (smaller MSE). Interestingly, in the high-dimensional setting, the packages \pkg{EMCluster} and \pkg{bgmm} exhibited worse performance. In particular, as can be seen in panel E of supplementary Figure 20, the proportions of the components recovered the $]0-1[^k$ simplex.


Conversely, the \pkg{EMCluster} package, and to a lesser extent, the \pkg{bgmm} package, performed surprisingly well when datasets were simulated with an underlying spherical covariance structure, even though the estimation was not performed explicitly with this constraint (Supplementary Table 19). Indeed, it seems like that the off-diagonal terms tended to converge towards 0, as showcased in Supplementary Figure 21, in Panel C, in which the fourth row from top represents the bootstrap intervals associated to the pairwise covariance between dimension 1 and 2 of each cluster.


**Unbalanced and overlapping components**

With unbalanced components and high $\operatorname{OVL}$ (scenario U9 in Table
\@ref(tab:parameter-configuration-univariate)), all packages, no
matter the initialisation method, provided biased estimates, with a higher variability of the parameter estimates compared to other configurations. The least biased
estimates were obtained with *k*-means or random initialisation, but with a higher variability on the estimates with random initialisation (Supplementary Table 9).
Delving further into the individual analysis of the parameter estimates associated to each component, we found out that the least biased estimates were achieved with rebmix initialisation for the most distinguishable components. For instance, in our configuration, the clusters 2 and 4 (see Supplementary Figure 7 and Table 9) were better characterised with the rebmix method. This observation aligns with the rebmix's underlying framework, using the modes of the distribution for initialising the component [@rebmix2015a]. With highly-overlapping distributions and unbalanced components, both the choice of the initialisation algorithm and the package have a
substantial impact on the quality of the estimation of this mixture.



Two configurations in our bivariate simulation feature distributions with both strong OVL and unbalanced components. Especially, the scenario B11 (Table \@ref(tab:parameter-configuration-bivariate)) has the strongest OVL overall, with notably a risk of wrongly assigning minor component 2 to major component 1 of 0.5 (a random method classifying each observation to cluster 1 or 2 would have the same performance).

First, we observe that the the random and rebmix initialisation methods have similar performance, significantly better than $k$-means or MBHC (Supplementary Figure 11). Specifically, the rebmix method returns the least biased estimates, while the random method is associated with the lowest MSE (respectively for configurations B11 and B15, the supplementary Tables 11
and 14). Second, the estimates differ across packages only in these two complex configurations, with packages \pkg{Rmixmod} and \pkg{mixtools} returning more accurate estimates than the others.
It it is particularly emphasised in Scenario B15, where the component-specific covariance matrices are diagonal with same non-null input, and thus should present spherical density distributions. However, only the first class of packages correctly recovers the spherical bivariate $95\%$ confidence regions while they are slightly ellipsoidal with the second class of packages (Panel B, Supplementary Figure 14).


With full covariance structures and unbalanced proportions, as depicted in the high-dimensional Scenario HD8a) and b) of Table \@ref(tab:parameter-configuration-HD), the general observations stated in the previous subsection for the high dimensional setting hold, namely that the least biased estimates are returned by packages not specifically designed for high-dimensional data, with the *k*-means initialisation (Supplementary Table 12 and supplementary Figure 22). Furthermore, the \pkg{EMCluster} and \pkg{bgmm} packages and the two packages dedicated to high-dimensional, perform similarly with $n=200$ observations (sub-scenario a) and $n=2000$ observations (sub-scenario b), whereas we would expect narrower and less biased confidence intervals by increasing the number of observations by a factor of 10.

Finally, with spherical covariance structures and unbalanced proportions, the best performances, both in terms of bias and variability, are obtained with \pkg{flexmix}, \pkg{mclust} and \pkg{GMKMCharlie}. Indeed, as detailed later in [Conclusions], these packages are more sensitive to the choice of the initialisation method and have a greater tendency to get trapped in the neighbourhood of the initial estimates (Supplementary Table 19 and supplementary Figure 22). Accordingly, *k*-means initialisation performs best since it assumes independent and homoscedastic features for each cluster.
Furthermore, \pkg{EMMIXmfa} is the package that best estimates the off-diagonal terms in this setting, as highlighted in supplementary Table 19.


**Identification of two classes of packages with distinct behaviours**

By summarizing the results obtained across all simulations, we identify two classes of packages with distinct behaviours (Figure
\@ref(fig:dichotomy-package-conclusion)):

* The first class of packages, represented by \pkg{Rmixmod} and \pkg{mixtools}, returns similar estimates to our baseline EM implementation. The estimates returned by these packages are less biased but at the extent of a higher variability on the estimates. Additionally, with overlapping mixtures, they tend to be slower compared to the second class, since they require additional steps to reach convergence.

* The second class of packages, composed of the other
reviewed packages, is more sensitive to the initialisation method. This leads to more
biased but less variable estimates, especially when assumptions done by
the initialisation algorithm are not met.

Panels A, B and C display, respectively in the univariate, bivariate and high-dimensional setting, the heatmap of the Pearson correlation between the estimated parameters across the benchmarked packages for the most discriminative scenario (the one featuring the most unbalanced and overlapping components): scenario U9, Table \@ref(tab:parameter-configuration-univariate) in the univariate setting, scenario B11, Table \@ref(tab:parameter-configuration-bivariate), for the bivariate simulation and scenario HD8, Table \@ref(tab:parameter-configuration-HD) for the high-dimensional simulation, with the *k*-means initialisation.

We further identified with this representation minor differences for the estimation of the parameters between \pkg{Rmixmod} and \pkg{mixtools}, while three subgroups can be identified in the second class of packages: the first subset with \pkg{bgmm} and \pkg{mclust}, the second subset with \pkg{EMCluster} and \pkg{GMKMcharlie} packages and the \pkg{flexmix} package, which clearly stands out from the others, as being the one most likely to be trapped at the boundaries of the parameter space. After examining the source codes of the packages, we attribute this differences to custom implementation choices of the EM algorithm, such as the way numerical underflow is managed or the choice of a relative or absolute scale to compare consecutive computed log-likelihoods (see Appendix *EM-implementation differences across reviewed packages* and Panel D, Figure \@ref(fig:dichotomy-package-conclusion)). In the high-dimensional setting, the second class of packages showed additional heterogeneity, with \pkg{EMCluster} and \pkg{bgmm} setting themselves apart from the other three packages.



```{r load-configurations}

univariate_configuration <- readRDS("./tables/univariate/univariate_configuration_scenario.rds")
bivariate_configuration <- readRDS("./tables/bivariate/bivariate_configuration_scenario.rds")
HD_configuration <- readRDS("./tables/HD/HD_configuration_scenario.rds")
```

```{r prepare-legend-final-figure}
html_ouput <- knitr::is_html_output()
html_final_legend <- paste("Panels A, B and C show respectively the heatmap of the Pearson correlation in the univariate, bivariate and high-dimensional framework between the parameters estimated by the packages, evaluated for the most discriminating and complex scenario. The correlation matrix was computed using the function", downlit::autolink("stats::cor"), "with option *complete* to remove any missing value related to a failed simulation, and the heatmap generated with the Bioconductor package *ComplexHeatmap*.
Panel D represents a tree summarising the main differences between the benchmarked packages, in terms of the EM implementation. They are discussed in more detail in Appendix *EM-implementation differences across reviewed packages*.")

pdf_final_legend <- "Panels A, B and C show respectively the heatmap of the Pearson correlation in the univariate, bivariate and high-dimensional framework between the parameters estimated by the packages, evaluated for the most discriminating and complex scenario. The correlation matrix was computed using the function \\code{stats::cor} with option \\textit{complete} to remove any missing value related to a failed simulation, and the heatmap generated with the Bioconductor package \\BIOpkg{ComplexHeatmap}. Panel D represents a tree summarising the main differences between the benchmarked packages, in terms of the EM implementation. They are discussed in more detail in Appendix \\textit{EM-implementation differences across reviewed packages}."
```

```{r dichotomy-package-conclusion, fig.cap=if (html_ouput) html_final_legend else pdf_final_legend}
knitr::include_graphics("./figures/dichotomy_package_conclusion.png")
```

**Failed estimations**

Finally, in some cases, neither the specific EM algorithm implemented by each package nor the initialisation method were able to return an estimate with the expected number of components, or converged to a degenerate distribution (e.g., with infinite or zero variances).  In that case, we considered the estimation as failed and accordingly we did not include it into the visualisations and the summary metric tables.

Most of the failed estimations occurred with the rebmix initialisation. Indeed, an updated version of the package forced the user to provide a set of possible positive integer values for the number of components, with at least a difference of two between the model with the most components and the model with the least components (we therefore set the parameter *cmax* to $k+1$ and *cmin* to $k-1$).In scenarios where the distributions associated with each cluster exhibit significant overlap, there is an increased risk of incorrectly estimating the number of components. This arises from the inherent difficulty of discerning the modes within the overall distribution. For instance, in the most complex scenario B11, characterized by strong overlap and imbalanced clusters (refer to Table \ref{tab:parameter-configuration-bivariate}), up to $20\%$ of initialisations were unsuccessful. Similarly, in the second most challenging scenario, B15, approximately $10\%$ of initializations failed against an averaged number of $4\%$ of the simulations exhibiting an inaccurate estimation of the number of components.



Removing errors proceeding from the initialisation method, only the \pkg{flexmix} package failed in returning an estimate matching the user criteria in some configurations of the univariate and  bivariate settings. In both cases, the strong assumption that any cluster with less than $5\%$ of the observations is irrelevant, results in trimming one or more components^[With a two-components mixture like our bivariate scenario, this even implies to consider an unimodal distribution of the dataset]. This strong agnostic constraint on component proportions led to failures in configurations featuring strongly overlapping clusters, with up to $20\%$ failed estimations with the random initialisation method in scenario B11 (Table \@ref(tab:parameter-configuration-bivariate)) and $80\%$ failed estimations in the univariate setting^[the gap proceeds from the stronger level of imbalance and the greater number of components] with the rebmix initialisation with scenario U9, Table \@ref(tab:parameter-configuration-univariate).



In a relatively high dimensional framework, as tested on our third benchmark (Table \@ref(tab:parameter-configuration-HD)), none of the algorithms that were initialised with the random method (`r knitr::asis_output(ifelse(knitr::is_html_output(), downlit::autolink("EMCluster::rand.EM()"),   '\\code{EMCluster::rand.EM()}'))`) converged successfully. Indeed, of the 16 configurations tested, the covariance returned during the initialisation was systematically non-positive definite for at least one of the components, violating the properties of covariance matrices.
Furthermore, an analysis of summary metrics in scenarios HD1 and HD8, reported in supplementary Tables 20 and 21, revealed a notably higher rate of failures when employing rebmix initialisation in conjunction with packages tailored for high dimensionality, such as \pkg{HDclassif}  \pkg{EMMIXmfa}. This discrepancy was in stark contrast to the more reliable and consistent initial estimates returned by $k$-means or hierarchical clustering.

Furthermore, as shown by the comparison of summary metrics with  $n=200$ and $n=2000$ observations in supplementary Tables 20 and 21, respectively for the simplest scenario HD1 and the most complex one HD8, the rebmix initialisation on the one hand, and the packages dedicated to high dimensionality or those of the second class of packages that show a particular behaviour, present much more failures than the *k*-means or hierarchical clustering initialisation.



# Conclusions

There are many packages that implement the EM algorithm for estimating the parameters of GMMs. But only few are regularly updated,
implement both the unconstrained univariate and multivariate GMM, and enable the user to provide its own initial estimates. Hence, among the 54 packages dealing with GMMs available on CRAN or Bioconductor repositories, we focused our review on 7 packages which implement all of these features. We believe that our in-depth review of the packages can help users to quickly find the best package for their clustering
pipeline and highlight limitations in the implementation of some
packages. Our benchmark covers a much broader range of configurations
than the previously-published studies [@nityasuddhi_bohning03;
@lourens_etal13; @leytham84; @xu_knight10], as we studied the impact of the level of overlap and the imbalance of the mixture proportions on the quality of the estimation.


Interestingly, the EM algorithm occasionally yields biased and inefficient estimates when the components overlap a lot, which agrees with the past literature [@lourens_etal13; @leytham84; @xu_knight10]. This appears to go counter to the theoretical results presented by @leytham84, which demonstrated the asymptotic consistency, unbiasedness, and efficiency of maximum likelihood estimates of GMMs. However, it's important to note that this theoretical demonstration relies on the definition of a "local" environment, necessitating the prior setting of boundaries within which the theorem's conditions are met (in other words, the definition of the *support*, which delineates the region where the initial values can be sampled from). It's not then surprising that the EM algorithm struggles in reaching the global maximum of the distribution in the presence of multiple local extremes.


When all components are well-separated or have a relatively small number of components (three or fewer), we found that the best
estimation (lowest MSE and bias) is reached with the latest initialisation method developed, namely the rebmix one. Notably, the global maximum is always properly found in our simulations with distinguishable components. Yet, with overlapping components, the least biased and variable estimates overall are obtained with *k*-means initialisation, enforcing the robustness of the method while the assumptions for using it are not met. 

On the contrary, with unbalanced and numerous components (above three), the quantiles initialisation leads to the most biased estimates while the rebmix initialisation induces the highest variability. Indeed, rebmix initialisation is not fit for highly overlapping mixtures, since one of its most restrictive assumption is that each generated interval of the empirical mixture distribution can be associated unambiguously to a component (see [Initialisation of the EM algorithm] and @rebmix2015a). 

Furthermore, rebmix is not particularly adjusted to deal with high-dimensional mixtures, displaying systematically poorer performance compared to other initialisation strategies, such as *k*-means or hierarchical clustering, as illustrated by the summary metrics listed in Appendix *Supplementary Figures and Tables in the HD simulation*. Higher risk of returning a sub-optimal extremum likely arises from the increased data sparsity in high dimensional datasets, which grows as the square root of the number of dimensions $\sqrt{D}$ ([Convergence of distance definitions](https://en.wikipedia.org/wiki/Curse_of_dimensionality#Distance_function)). Thus, we expect that most of the equally-large intervals binning the sampling space and that are used to initiate the rebmix algorithm contain either no or only observation, preventing from retrieving the numerically defined mode of the distribution and increasing the risk of initiating the algorithm in a spurious neighbourhood.

About the remaining initialisation strategies, we observed that, even in the well-separated case, random initializations can sometimes yield highly biased estimates, far from the true parameter values. Consistent with our observations, it was shown in @jin_etal16 that the probability for the EM algorithm to converge from randomly initialised estimates to a local suboptimal maximum is non null above two components, increasing with the number of components. Additionally, the local maximum of the likelihood function obtained can be arbitrarily worse than the global maximum. Finally, hierarchical clustering does not take into account any uncertainty on the assignment for an observation to a given class, which explains its rather bad performances with overlapping components. Overall, there is always an initialisation algorithm performing better than the hierarchical clustering, and further it is also by far the slowest and most computationally intensive initialisation method (see  supplementary Figure 10).


Our study reveals that while the EM algorithm is supposed to be deterministic, the estimates obtained from its implementations can differ across packages. We were able to link these differences with custom choices of EM implementations across the benchmarked packages.
Two distinct classes of packages emerge, each with specific approaches to address certain limitations of the EM algorithm. The first class, exemplified by \pkg{mixtools} and \pkg{Rmixmod} typically yields smaller but less biased estimates that exhibit lower sensitivity to the choice of initialization method. However, these estimates tend to have higher variability and require longer running times to achieve convergence. The second class, composed of the remaining packages, provide estimates with reduced MSE, but at the extent of a higher bias on the estimates. One plausible explanation is that the first class of packages, comparing absolute iterations of the function to be maximised, tends on average to perform more iterations. The estimated results are accordingly more consistent and closer to the true MLE estimation but at the expense of an increased risk of getting trapped in a local extrema or a plateau, explaining the greater number of outliers observed. Among them, \pkg{flexmix} stands out by choosing
an unbiased but non MLE-estimate of the covariance matrix, without any clear improvement of the overall performance in our simulations.


Based on these results, we design a decision tree indicating the best choice of package and initialisation method in relation with the shape of the distribution, displayed in Figure \@ref(fig:decision-tree-GMMs). Interestingly, our conclusions are consistent between the univariate and bivariate settings. Furthermore, most of the general recommendations on the best choices of packages with respect to the characteristics of the mixture model generally hold in a relatively higher dimensional setting^[We should note, however, that a larger sample space revealed that the packages \pkg{bgmm} and \pkg{EMCluster} display more biased and noisy parameters compared to the other packages benchmarked and that their performance was surprisingly unaffected by the number of simulated realisations]. From this, we assume that projection into a lower-dimensional space is only beneficial in a very high-dimensional setting, for example when the number of dimensions exceeds the number of observations, or when unrestricted parameter estimation (with the full covariance structure) is practically infeasible for computational reasons.
<!-- Finally,  Notably, just like the packages dedicated to high dimensional setting, the inferred confidence intervals from bootstrap simulations disprove the expected predictions of the multivariate version of the Central -->
<!-- Limit Theorem (CLT)^[Indeed, the multidimensional CLT states that the empirical mean of a vector of unbiased parameters, each inferred independently (such the bootstrap distribution we estimated using the EM algorithm with a bootstrap strategy, expected theoretically to be asymptomatically unbiased) converges to a multivariate normal distribution, whose covariance matrix is proportional to the inverse of the square root of the number of observations. Notably, we would expect narrower confidence bands and less biased estimations by increasing the number of observations.].  -->




```{r decision-tree-GMMs, fig.cap="A decision tree to select the best combination of package and initialisation method with respect to the main characteristics of the mixture. It's worth pointing that in both univariate and low dimension multivariate settings, the recommandations are similar."}
# decision_tree_univariate <- Node$new("Estimation of the MLE in GMMs")
# separated <- decision_tree_univariate$AddChild("Well-separated components")
# rebmix <- separated$AddChild("Rebmix initialisation")
# overlapping <- decision_tree_univariate$AddChild("Overlapping components")
# balanced <- overlapping$AddChild("Balanced mixture")
# kmeans_balanced <- balanced$AddChild("*k*-means initialisation with \n GMKMcharlie or EMCluster")
# unbalanced_numerous_components <- overlapping$AddChild("Unbalanced and number of clusters \U2265 4")
# unbalanced_variability <- unbalanced_numerous_components$AddChild("Optimise variability and MSE")
# second_class <- unbalanced_variability$AddChild("Second class of packages \n with *k*-means initialisation, \n preferentially with GMKMcharlie \n or EMCluster")
# unbalanced_bias <- unbalanced_numerous_components$AddChild("Optimise bias")
# first_class <- unbalanced_bias$AddChild("First class of packages \n with *k*-means initialisation \n for proportions and variability \n estimations and random \n for centroids estimation")
# unbalanced_few_components <- overlapping$AddChild("Unbalanced and number of clusters < 4")
# decision_unbalanced_few_components <- unbalanced_few_components$AddChild("rebmix or random initialisation \n with the second class of packages.")
#
# # custom the decision tree plot
# SetGraphStyle(decision_tree_univariate, rankdir = "TB", fontSize = 25)
# SetEdgeStyle(decision_tree_univariate, arrowhead = "vee", penwidth = 2)
# SetNodeStyle(decision_tree_univariate, style = "filled,rounded", shape = "box",
#              fontname = "helvetica", tooltip = GetDefaultTooltip)
# # SetNodeStyle(separated, fillcolor = "chartreuse1")
# # SetNodeStyle(overlapping, fillcolor = "crimson")
# Do(decision_tree_univariate$leaves, function(node) SetNodeStyle(node, shape = "egg")) # custom the leaves
#
#
# DiagrammeR::export_graph(ToDiagrammeRGraph(decision_tree_univariate),
#                          "./figures/decision_tree.png", file_type = "png")


knitr::include_graphics("./figures/decision_tree_GMMs.png")
```


Comparing all these packages suggest several improvements.

1. The use of C++ code speeds up the convergence of the EM algorithm
compared to a native R implementation.

2. All packages dealing with GMMs should use *k*-means for overlapping, complex mixtures and rebmix initialisation for well-separated components, provided that the dimension of the dataset is relatively small. The final choice between these two could be set after a first run or visual inspection aiming at determining roughly the level of entropy across mixture proportions and the degree of overlap between components.

3. The packages should allow the user to set their own termination criteria (either relative or absolute log-likelihood or over the estimates after normalisation). Interestingly, \pkg{EMMIXmfa} is the only package among those examined that allows the user to consider an absolute or relative convergence endpoint of the EM algorithm, through the `conv_measure` attribute, with `diff` and `ratio` options respectively.

4. With a great number of components or complex overlapping distributions, the optimal package should integrate prior information when available, e.g. via Bayesian estimation.


While \pkg{mclust} appeared as the most complete package to model GMMs in R, none of the packages reviewed in this report features all the characteristics mentioned above.
We thus strongly believe that our observations will help users identify the most suitable packages and parameters for their analyses and guide the development or updates of future packages.

# Simulation settings
For ease of reading, we reproduce below the parameter configurations used to run the three benchmarks, respectively for the univariate (Table \@ref(tab:parameter-configuration-univariate)), bivariate (\@ref(tab:parameter-configuration-bivariate)) and high dimensional setting (Table \@ref(tab:parameter-configuration-HD)).

```{r parameter-configuration-univariate, layout = "l-body-outset"}
univariate_configuration %>%
  mutate(
    Proportions = purrr::map_chr(true_parameters, ~ paste(.x$p, collapse = " / ")),
    Means = purrr::map_chr(true_parameters, ~ paste(.x$mu, collapse = " / ")),
    Correlations = purrr::map_chr(true_parameters, ~ paste(.x$sigma, collapse = " / "))
  ) %>%
  select(-c("OVL", "prop_outliers", "true_parameters", "formatted_true_parameters", "old_OVL")) %>%
  rename(Entropy = entropy, OVL = OVL_pairwise) %>%
  kbl(
    booktabs = T, caption = "The 9 parameter configurations tested to generate the samples of the univariate experiment, with $k=4$ components.",
    escape = F, align = "c"
  ) %>%
  kable_styling(latex_options = c("hold_position", "scale_down")) %>%
  row_spec(0, bold = T) %>%
  row_spec(1:9, hline_after = T) %>%
  kable_styling(bootstrap_options = c("hover", "condensed"))
```

```{r parameter-configuration-bivariate, layout = "l-body-outset"}
bivariate_configuration %>%
  mutate(
    Proportions = purrr::map_chr(true_parameters, ~ paste(.x$p, collapse = " / ")),
    Means = purrr::map_chr(true_parameters, ~ paste0("(", paste0(.x$mu[, 1], collapse = ","), ");(", paste0(.x$mu[, 2], collapse = ","), ")")),
    Correlations = purrr::map_chr(true_parameters, ~ paste(c(.x$sigma[1, 2, 1], .x$sigma[1, 2, 2]), collapse = " / "))
  ) %>%
  select(-c("OVL", "nobservations", "prop_outliers", "true_parameters", "formatted_true_parameters", "true_parameters_factor")) %>%
  rename(Entropy = entropy, OVL = OVL_pairwise) %>%
  kbl(
    booktabs = T, caption = "The 20 parameter configurations tested to generate the samples of the bivariate experiment.",
    escape = F, align = "c"
  ) %>%
  kable_styling(latex_options = c("hold_position", "scale_down")) %>%
  row_spec(0, bold = T) %>%
  row_spec(1:20, hline_after = T) %>%
  kable_styling(bootstrap_options = c("hover", "condensed"))
```

```{r parameter-configuration-HD, layout = "l-body-outset"}
HD_configuration_formatted <- HD_configuration %>%
  rename(Entropy = entropy, Spherical = is_diagonal, `Number of observations` = nobservations) %>%
  mutate(
    Proportions = purrr::map_chr(true_parameters, ~ paste(round(.x$p, digits = 2), collapse = " / ")),
    Spherical = dplyr::if_else(Spherical, "\\greentick", "\\redcross"), ID = paste0("HD", ID)
  ) %>%
  select(-c("true_parameters", "Entropy")) %>%
  dplyr::relocate(Spherical, .after = Proportions)


HD_configuration_formatted %>%
  kbl(
    booktabs = T, caption = "The 16 parameter configurations tested to generate the samples in a high dimensional context. The first digit of each ID index refers
      to an unique parameter configuration (identified by its level of overlap, entropy and topological structure, either circular or ellipsoidal,
      of the covariance matrix, while the lowercase letter depicts the number of observations, a) with $n=200$ and b) with $n=2000$.",
    escape = F, align = "c",
    col.names = linebreak(c("ID", "OVL", "Number of \nobservations", "Proportions", "Spherical"))
  ) %>%
  kable_styling(latex_options = c("hold_position", "scale_down")) %>%
  row_spec(0, bold = T) %>%
  row_spec(1:16, hline_after = T) %>%
  kable_styling(bootstrap_options = c("hover", "condensed"))
```

# Additional files

* Additional files related to the univariate setting

    + **S1.** Bootstrap distributions of the estimated parameters for each
    scenario described in \@ref(tab:parameter-configuration-univariate).
    + **S2.** Mean, standard deviation, bias and MSE for each individually
    estimated parameter in configurations listed in \@ref(tab:parameter-configuration-univariate).
    + **S3.** Distribution of the running times taken for the EM estimation of the parameters of the GMM, across all nine configurations described in \@ref(tab:parameter-configuration-univariate), for each benchmarked package. We selected the *k*-means algorithm to initialise the EM algorithm, as being the least variable for a given package and scenario.
    + **S4.** Distribution of the time computations taken by the six initialisation methods listed in Table `r knitr::asis_output(ifelse(knitr::is_html_output(), '\\@ref(tab:general-parameter-description-html)', '\\@ref(tab:general-parameter-description-pdf)'))`.

* Additional files related to the outliers setting:

    +  **S5.** Bootstrap distributions of the estimated parameters used to
    generate Supplementary Figure 2. We additionally include the \pkg{otrimle} package, dedicated to these extreme distributions. Two configurations were tested, introducing $2 \%$ and $4 \%$ of outliers drawn from an improper uniform distribution.
    + **S6.** Mean, standard deviation, bias and MSE for each individually
    estimated parameter in both configurations visualised on Supplementary Figure 2, for each combination of package and
    initialisation method.


* Additional files related to the bivariate benchmark:

    + **S7.** Bootstrap distributions of the estimated parameters for each
    scenario described in \@ref(tab:parameter-configuration-bivariate).

    + **S8.** Mean, standard deviation, bias and MSE for each individually
    estimated parameter in configurations listed in \@ref(tab:parameter-configuration-bivariate).

    + **S9.** Distribution of the running times taken for the EM estimation of the parameters of the GMM, across all twenty configurations described in \@ref(tab:parameter-configuration-bivariate), for each benchmarked package. We selected the *k*-means algorithm to initialise the EM algorithm, as being the least variable for a given package and scenario.

* Additional files related to the high-dimensional benchmark:

    + **S10.** Bootstrap distributions of the estimated parameters for each
    scenario described in \@ref(tab:parameter-configuration-HD).

    + **S11.** Mean, standard deviation, bias and MSE for each individually
    estimated parameter in configurations listed in \@ref(tab:parameter-configuration-HD).

    + **S12.** Distribution of the running times taken for the EM estimation of the parameters of the GMM, across all twenty configurations described in \@ref(tab:parameter-configuration-HD), for each benchmarked package. We selected the *k*-means algorithm to initialise the EM algorithm, as being the least variable for a given package and scenario.

